singularity-eos/eos/eos_vinet.hpp

```diff
@@ -0,0 +1,377 @@
+      const Real rho, const Real temp, Real *lambda = nullptr) const;
+  PORTABLE_INLINE_FUNCTION Real BulkModulusFromDensityTemperature(
+      const Real rho, const Real temp, Real *lambda = nullptr) const;
+  PORTABLE_INLINE_FUNCTION Real BulkModulusFromDensityInternalEnergy(
+      const Real rho, const Real sie, Real *lambda = nullptr) const;
+  // Thermal expansion coefficient added AEM 2022
+  PORTABLE_INLINE_FUNCTION Real TExpansionCoeffFromDensityTemperature(
+      const Real rho, const Real temp, Real *lambda = nullptr) const;
+  PORTABLE_INLINE_FUNCTION Real GruneisenParamFromDensityTemperature(
+      const Real rho, const Real temp, Real *lambda = nullptr) const {
+    return _A0 * _B0 / _Cv0 / rho;
+  }
+  PORTABLE_INLINE_FUNCTION Real GruneisenParamFromDensityInternalEnergy(
+      const Real rho, const Real sie, Real *lambda = nullptr) const {
+    return _A0 * _B0 / _Cv0 / rho;
+  }
+  PORTABLE_INLINE_FUNCTION void FillEos(Real &rho, Real &temp, Real &energy, Real &press,
+                                        Real &cv, Real &bmod, const unsigned long output,
+                                        Real *lambda = nullptr) const;
+  PORTABLE_INLINE_FUNCTION
+  void ValuesAtReferenceState(Real &rho, Real &temp, Real &sie, Real &press, Real &cv,
+                              Real &bmod, Real &dpde, Real &dvdt,
+                              Real *lambda = nullptr) const;
+  // Generic functions provided by the base class. These contain e.g. the vector
+  // overloads that use the scalar versions declared here
+  SG_ADD_BASE_CLASS_USINGS(Vinet)
+  PORTABLE_INLINE_FUNCTION
+  int nlambda() const noexcept { return 0; }
+  static constexpr unsigned long PreferredInput() { return _preferred_input; }
+  PORTABLE_INLINE_FUNCTION void PrintParams() const {
+    static constexpr char st[]{"Vinet Params: "};
+    printf("%s rho0:%e T0:%e B0:%e BP0:%e\n  A0:%e Cv0:%e E0:%e S0:%e\n", st, _rho0, _T0,
+           _B0, _BP0, _A0, _Cv0, _E0, _S0);
+    for (int i = 0; i < 39; i++) {
+      if (_d2tod40[i] > 0.0) printf("d%i:%e\t", i + 2, _d2tod40[i]);
+    }
+    printf("\n\n");
+  }
+  // Density/Energy from P/T not unique, if used will give error
+  PORTABLE_INLINE_FUNCTION void
+  DensityEnergyFromPressureTemperature(const Real press, const Real temp, Real *lambda,
+                                       Real &rho, Real &sie) const;
+  inline void Finalize() {}
+  static std::string EosType() { return std::string("Vinet"); }
+  static std::string EosPyType() { return EosType(); }
+
+ private:
+  static constexpr const unsigned long _preferred_input =
+      thermalqs::density | thermalqs::temperature;
+  Real _rho0, _T0, _B0, _BP0, _A0, _Cv0, _E0, _S0;
+  static constexpr const int PressureCoeffsd2tod40Size = 39;
+  static constexpr const int VinetInternalParametersSize = PressureCoeffsd2tod40Size + 4;
+  Real _VIP[VinetInternalParametersSize], _d2tod40[PressureCoeffsd2tod40Size];
```

Comment:
I don't believe these will copy into GPU executed lambdas. I recall having this discussion somewhere but I forget where. This will need to be changed to some sort of other storage mechanism along with non-trivial getondevice and finalize functions.

Reply:
@dholladay00 would the running the test that @aematts already has made verify this if it's run on e.g. a volta-x86 node of Darwin? I would recommend changing one of the `d2to40` values to be non-zero for a sanity check.

Reply:
My concern is that it will fail to compile for GPUs. I know if you do:
```c++
Real arr[20];
Kokkos::parallel_for(N, KOKKOS_LAMBDA(int const& i) { x = f(arr); });
```
`nvcc` will complain about capturing that array. I haven't tested, but would assume that capturing a struct that contains one of those would also result in the same compilation failure but I haven't explicitly tested that.

Reply:
I haven't played around with it on GPUs, but I _believe_ a struct/class that contains an array member with known length will trivially copy:
https://godbolt.org/z/qevzxs577

Again, I haven't played around with this trying to copy to device, but I thought I remembered hearing that an object being trivially copyable was enough.

Reply:
I don't disagree WRT trivially copyable, but I recall these arrays are explicitly forbidden from being captured into a device lambda, as in nvcc will complain. 

I'll try to get a reproducer going.

Reply:
In the meantime, @aematts, can you try compiling and testing on Darwin if you have an account there? I'll e-mail you a build recipe tomorrow.

Reply:
After a discussion over webx, I think we stick with the static arrays for now. It should work on GPU with the `--expt-relaxed-constexpr` flag, which is already in place.